Fm refactor

fmeier opened the pull request:
draft refactor - please comment :)

exhaustin commented on differentiable_robot_model/data_utils.py:
```diff
@@ -10,6 +10,9 @@
 from torch.utils.data.dataset import Dataset
 
 
+torch.set_default_tensor_type(torch.FloatTensor)
```
Are we allowing users to set default tensor type for GPU-based experiments? If so we shouldn't be setting the default tensor type here.

fmeier replied:
oh I think this was a left over line, the code wasn't working on Mac anymore after the GPU PR, but then your new test PR fixed things (which I merged into my branch), and I forgot to remove my "fixes"

exhaustin commented on differentiable_robot_model/rigid_body_params.py:
```diff
@@ -38,38 +38,40 @@ def __init__(
         is_using_positive_initial_guest=False,
         init_param=None,
         is_initializing_params=True,
+        device='cpu'
```
For `torch.nn.Modules`, it would be cleaner to call `to()` outside the module after initialization to set the device for all parameters within the module and its submodules, instead of having to specify the device at every layer.

fmeier replied:
I had tried that initially but somehow it kept failing, I'll take a look again

exhaustin reviewed (commented):
Regarding the two types of parametrization: constants & nets:
- I don't understand why we need to differentiate between the two types, given that `UnconstrainedMassValue` behaves similar to a torch.nn.Parameter. What would happen if we change all of the learnable params to modules, and thus we can omit the "module" field in `learnable_params[<param>]`? This way users also gain the freedom to parametrize masses as a single param, or inject structure/constraints into translation & rotation.
- If we want to keep the two types, the "module" key still seems extraneous since the content can already be determined from the parameter key ("mass", "trans"...).

One big concern I have is whether we will want to support different parameterizations for each link. For example, users might want to learn the mass for one link but the trans for another link, or users might want to parametrize masses of different links with different models. If we want to support this in the foreseeable future, then I think it's best to either support it now or build the API in a way that's extendable in that direction, else that's another major refactor down the road.

One idea I have is to have an interface for setting parameterizations of the robot model dynamically like such: `robot_model.set_parameterization(link_name, property, parametrization, init_value)`. For modules, this will look like `robot_model.set_parameterization("iiwa_link_1", "mass", UnconstrainedMassValue, None)`, and for params we can have `robot_model.set_parameterization("iiwa_link_2", "trans", None, 0.1)`. We can even create macros that wrap around this API such as `robot_model.set_parameterization_for_all_links("mass", UnconstrainedMassValue, params)`, or `robot_model.set_parameterization_from_dict(cfg)`.

fmeier commented:
I like the API change idea! should I do it myself, or do you want to do it @exhaustin ?

exhaustin commented:
> I like the API change idea! should I do it myself, or do you want to do it @exhaustin ?

Glad you liked it!!
I can go through and fix the device issues for you since I just went through a related PR and thus the details are still in the L1 cache of my brain.
I am currently limited in terms of bandwidth, so I might not have time to do the API refactor before the end of next week. Maybe we can talk about it on Monday?

fmeier commented:
oh what I meant to say is that after merging with master (which had your latest PR) the device issue was fixed - just have to remove that line again :) so nothing to do here

I'll make a first stab at the API change, then we can look at it together on Monday :)

fmeier commented:
ah ok, yes somehow when I tried the .to(device) now, it just worked.
I pushed an initial version of the suggested API (only the simple version) - take a look, and let me know what you think. Have a good weekend :)

fmeier commented:
I modified your unlearnable functionality into freeze_link_param, and then also added unfreeze_link_param -> I think that is all we need

also on Mac type(param_module) did not equal torch.nn.Module so I modified that a bit -> can we add circle ci testing for Mac?

let me know what you think.

exhaustin commented:
I like the freeze/unfreeze API! 

Regarding tests for different parametrizations, I think we just need to check that they produce outputs in the correct format and are trainable when `forward` is being called right? If that's ok I can try adding some unit tests.

I think I can try to add tests for mac this week. Just need to figure out how to locate an image and also disable the gpu tests.
In the meantime this PR looks good to merge!
